dora/core/common/src/main/java/alluxio/membership/AlluxioEtcdClient.java

```diff
@@ -0,0 +1,557 @@
+                    GetOption.newBuilder().isPrefix(true).build())
+                .get(DEFAULT_TIMEOUT_IN_SEC, TimeUnit.SECONDS);
+            return getResponse.getKvs();
+          }, new ExponentialBackoffRetry(RETRY_SLEEP_IN_MS, MAX_RETRY_SLEEP_IN_MS, RETRY_TIMES));
+    } catch (AlluxioRuntimeException ex) {
+      throw new IOException(ex.getMessage());
+    }
+  }
+
+  /**
+   * Add listener to a path internal function.
+   * @param parentPath
+   * @param listener
+   * @param watchType
+   */
+  private void addListenerInternal(
+      String parentPath, StateListener listener, WatchType watchType) {
+    if (mRegisteredWatchers.containsKey(getRegisterWatcherKey(parentPath, watchType))) {
+      LOG.info("Watcher already there for path:{} for children.", parentPath);
+      return;
+    }
+    WatchOption.Builder watchOptBuilder = WatchOption.newBuilder();
+    switch (watchType) {
+      /* e.g. Given the parentPath '/parent/',
+      give query-like syntax equivalent to:
+        select * with value < '/parent.' ('.' the char before '/' in ASCII)
+      which includes all keys prefixed with '/parent/' */
+      case CHILDREN:
+        String keyRangeEnd = parentPath.substring(0, parentPath.length() - 1)
+            + (char) (parentPath.charAt(parentPath.length() - 1) + 1);
+        watchOptBuilder.isPrefix(true)
+            .withRange(ByteSequence.from(keyRangeEnd, StandardCharsets.UTF_8));
+        break;
+      case SINGLE_PATH:
+      default:
+        break;
+    }
+
+    Watch.Watcher watcher = mClient.getWatchClient().watch(
+        ByteSequence.from(parentPath, StandardCharsets.UTF_8),
+        watchOptBuilder.build(),
+        new Watch.Listener() {
+          @Override
+          public void onNext(WatchResponse response) {
+            for (WatchEvent event : response.getEvents()) {
+              switch (event.getEventType()) {
+                case PUT:
+                  listener.onNewPut(
+                      event.getKeyValue().getKey().toString(StandardCharsets.UTF_8),
+                      event.getKeyValue().getValue().getBytes());
+                  break;
+                case DELETE:
+                  listener.onNewDelete(
+                      event.getKeyValue().getKey().toString(StandardCharsets.UTF_8));
+                  break;
+                case UNRECOGNIZED:
+                default:
+                  LOG.info("Unrecognized event on watch path of:{}", parentPath);
+                  break;
+              }
+            }
+          }
+
+          @Override
+          public void onError(Throwable throwable) {
+            LOG.warn("Error occurred on children watch for path:{}, removing the watch.",
+                parentPath, throwable);
+            removeChildrenListener(parentPath);
+          }
+
+          @Override
+          public void onCompleted() {
+            LOG.warn("Watch for path onCompleted:{}, removing the watch.", parentPath);
```

Comment:
nit: WARN or INFO?

Reply:
it should be WARN as if we got onCompleted callback which means this lease is revoked, mostly in unwanted cases as we don't proactively revoke them